# Tutorial 3: Variables and Strings

## PHYS 2600, Spring 2019

## T3.1 - Using variables and assignment

Let's begin with some practice using variables and the assignment operator `=`.


### Part A

Try to run the cell below - __why doesn't it work?__  Fix the code so that the cell runs and prints the answer (5).

In [1]:
### BEGIN SOLUTION
# This cell fails because of how assignment works in Python!
# We can't save the symbolic expression "A + B" to C, instead
# Python tries to evaluate "A + B" and finds those variables don't exist.

A = 2
B = 3
C = A + B

### END SOLUTION

C = A + B
A = 2
B = 3
print(C)



5


### Part B

The cell below defines variables `pi` and `r`, containing the constant $\pi$ and the radius of a circle $r$.  Use the variables to calculate the area of a circle with the given radius of 2 meters.  __Save your answer to the variable `area`.__

In [2]:
pi = 3.14159
r = 2

### BEGIN SOLUTION
area = pi * r**2
### END SOLUTION


Once you've finished the cell above, run the below __testing cell__.  If you did everything right, the cell below will _run and produce no output_: this means your code passes the test.  On the other hand, if there is a problem, you will see an _error message_ when you run the cell below.

(These testing cells will be very common in tutorial and homework, as ways to check that your code is working correctly.)

In [3]:
## Testing cell - read the Markdown cell above, then run me!
# Don't worry about the syntax here yet - you'll learn to write your own tests soon.

assert (area - 12.566) <= 0.001

Now we'd like to find the area of a circle with a radius of 5 meters instead.  _Why doesn't the cell below work as written?_  Once you think you know the problem, __modify the code to print the right area__ (the answer should be about 78.5.)

(If you're not sure what's going wrong, copy your code from the first cell and from the one below into the [Python Tutor](http://www.pythontutor.com/visualize.html#mode=edit) and see what happens line by line.)

In [4]:
r = 5  # Update r to 5 meters

### BEGIN SOLUTION
# Remember how assignment works: when we write area = pi * r**2,
# area gets the _numeric value_ from evaluating the right-hand side.
# Since it doesn't store a symbolic expression, changing r later
# won't change area!

area = pi * r**2
### END SOLUTION

print(area)

78.53975


## T3.2 - String formatting

Use `%` to create the following as a formatted string, using the two variables set below:

`40% of all people know that 82.4% of all statistics are made up on the spot.`

In [6]:
people_know = 40
made_up = 82.4

### BEGIN SOLUTION
print('%d%% of all people know that %g%% of all statistics are made up on the spot.' % (people_know, made_up))
### END SOLUTION


40% of all people know that 82.4% of all statistics are made up on the spot.


Write the following (with each color on a new line!) from a `print` statement on a single line by using the `\n` character:

`Red
Green
Blue`

In [7]:
print("Red\nGreen\nBlue")

Red
Green
Blue


There is a way to do the last example, i.e. to save the list of colors with line breaks to a single string, _without_ using the newline character `\n`.  Can you do it in the cell below?

In [8]:
# Using a triple-quoted string lets us not use \n explicitly!
print("""Red
Green
Blue""")

Red
Green
Blue


Using a single `print()` statement in the code box below, produce the following string as output:

`There's no 'I' in "team"!`

_(Hint: to do this with one `print()`, you'll either need to use the escape character `\`, or you'll need to concatenate two strings together with `+`.)_

In [9]:
# First option: escape characters
my_str = "There's no 'I' in \"team\"!"
print(my_str)

# Second option: string concatenation
first_str = "There's no 'I' "
second_str = 'in "team"!'
print(first_str + second_str)

There's no 'I' in "team"!
There's no 'I' in "team"!


## T3.3 - An algorithm for the square root

The machine-code instructions a CPU can execute are limited to fairly simple mathematical and logical operations; for anything more complicated, a human being needs to implement an _algorithm_ to carry out the calculation or task we want.

Let's walk through a simple algorithm for computing the __square root__ of a number, using only simple math operations and some logic.  (In fact, modern CPUs can take a square root as a basic instruction, but many older ones could not, so they would have had to use something like this.)

Here is the algorithm we will use, in words, for computing $y = \sqrt{x}$:

1. Start with an initial guess for y.
2. Calculate y*y, and stop if it's "close enough" to x.
3. Update y to the average of y and x/y: y --> (y+x/y)/2.
4. Loop back to step 2.

(This algorithm actually [dates back to the ancient Greeks](https://en.wikipedia.org/wiki/Methods_of_computing_square_roots)!)

This is an example of an __iterative algorithm__, which is very common in computer programming.  It relies on a repeated sequence of steps to find the answer to a problem; the "loop back" step 4 above provides the repetition.

At every step, we only need to know the current value of `y` to find the next one.  Since we don't need the full history of `y`, we can carry out the algorithm by __updating__ `y` at every step and overwriting the previous value.

(In this context, we might call `y` a __temporary variable__, a __register variable__, or a __placeholder variable__.  All of these terms convey the idea that we re-use the name `y`, and it holds only the value from the last iteration we did.)

### Part A

Before we get to the square root, let's see the basic idea of variable updating.  Modify the cell below __using only copy and paste__ until the value 30 is stored in the variable `x`.

In [11]:
x = 0  ## Initial value for x

x = x + 5
### BEGIN SOLUTION
x = x + 5
x = x + 5
x = x + 5
x = x + 5
x = x + 5
### END SOLUTION

print(x)

30


In [ ]:
## Another testing cell!
## If you did the cell above correctly, this cell will run and do nothing.
## Otherwise, it will give an error message!

assert x == 30

### Part B

Now let's use Python to find the square root of 749.  In the cell below, __implement the square root algorithm above__.  Each time you reach step 2, use a `print()` statement to check the value of `y*y` and see whether you need to add another loop.  Once again, using copy/paste is recommended here!

Take your initial guess to be y=20, and let's say "close enough" is when our estimate for y*y is within 0.2 of 749.  (If you're doing it right, this should converge pretty quickly!)

In [12]:
y=20
x=749

### BEGIN SOLUTION
print(y*y)
y=(y+x/y)/2
print(y*y)
y=(y+x/y)/2
print(y*y)
y=(y+x/y)/2
print(y*y)
print(y)
### END SOLUTION

400
825.1256250000001
750.7558268117176
749.0010266080137
27.367883122521803


The correct answer for the square root (as per my calculator) is 27.3679 - notice that we did very well, and the algorithm converged pretty quickly!  

Other methods to compute the square root may have better (or worse) convergence, i.e. may require less (or more) steps to get the answer to the precision we want.

Some final thoughts to consider, if you've reached the end early:

* What happens if you start with a negative number as your guess for y?  Does the result make sense?
* How much does the precision of your result improve as you keep adding more iterations of the algorithm?